### DISCUSSION TITLE
i.colors.enhance for single bands?

### DISCUSSION THREAD

**Discussion by CarlosGrohmann** [OSGeo/grass#1942] [Q&A]
Hello all

Today as I was running some analysis, I needed to do a histogram stretch for a single raster. The thing is, `i.colors.enhance` is made to run with 3 bands, in order to create a nice RGB composition. 

So I went and modified it to run with a singe band.

The question is:do you think a module like `i.colors.enhance.single` would be a good contribution? Maybe with a different name, such as `r.histo.stretch` as it can be used in any raster.

If so, how to proceed? Add as an addon?

Thanks

**Comment by veroandreo** [comment]
There's r.colors for single raster maps and it has histogram equalization flag. That does not work for you?

**Reply by CarlosGrohmann** [reply]
No, Vero. The idea is to achieve histogram stretching (removing 1-2% of the tails), not equalization.

**Reply by veroandreo** [reply]
I see. I do not know what others think but instead of a new module, I'd suggest to add a new option to r.colors for histogram stretching. What do you think?

**Reply by CarlosGrohmann** [reply]
I can't make changes to `r.colors` because it is in C. `i.colors.enhance` is in python so I can adapt/change. Another option is to add some optional features to `i.colors.enhance` such as running it on a single band, multiple bands, or an image group.

**Reply by ninsbl** [reply]
`r.colors` works with percentages in color rules directly.

So, you could create a rules file with:
```
0% black
2% black
98% white
100% white
```

Or you use absolte values with e.g. percentiles derived from `r.univar`.
Simplifying this with a wrapper module can make it easier for users. Like @veroandreo, I would also prefer having this as an option in `r.colors` directly, to clamp the color values to some user defined percentiles. Yet, I see that `r.colors` is C as you rightly point out, so I could not work on that either...
But having this i tight to `r.colors`, as e.g. `r.colors.stretch` could probably allow to apply this functionality to various color tables and not just greyscale?! But that would require to read the color table and then modyfy it in Python before writing it to the raster...

**Reply by neteler** [reply]
> `r.colors` works with percentages in color rules directly.
...
> Or you use absolte values with e.g. percentiles derived from `r.univar`. 

Note that the original script, to remove outliers uses percentiles as in area under the histogram curve. See

https://github.com/OSGeo/grass/blob/dfeeadb48a626e323eac3286ce7b4c621dcc941a/scripts/i.colors.enhance/i.colors.enhance.py#L76

which is based on [r.quantile](https://grass.osgeo.org/grass-stable/manuals/r.quantile.html) (_"r.quantile computes quantiles in a manner suitable for use with large amounts of data. It is using two passes."_), not `r.univar`. Maybe it is the same as you propose, I cannot check at time.
